### DISCUSSION TITLE
How to learn about new packages in homebrew?

### DISCUSSION THREAD

**Discussion by mathomp4** [Homebrew/discussions#4337] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.0.6-147-gb3684e5
ORIGIN: https://github.com/Homebrew/brew
HEAD: b3684e55a7a0090e50d975c347961cb57a5ffcb5
Last commit: 26 hours ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 0c46c555dcb43e7ae3147392a71ebc2c1f5e0a26
Core tap last commit: 5 weeks ago
Core tap branch: master
Core tap JSON: 18 Mar 21:46 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.P6enDcbzX3/org.xquartz:0
HOMEBREW_EDITOR: /usr/local/bin/vim
HOMEBREW_GITHUB_API_TOKEN: set
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 14.0.0 build 1400
Git: 2.40.0 => /usr/local/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2.1-x86_64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2
```

### Output of `brew doctor`

```shell
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  python@2
  pycparser
  libgit2@1.5
  grv
  erdtree
  unrar

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  ilmbase
  jsonschema

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/fakemysql.h
  /usr/local/include/fakepq.h
  /usr/local/include/fakesql.h
  /usr/local/include/itcl.h
  /usr/local/include/itcl2TclOO.h
  /usr/local/include/itclDecls.h
  /usr/local/include/itclInt.h
  /usr/local/include/itclIntDecls.h
  /usr/local/include/itclMigrate2TclCore.h
  /usr/local/include/itclTclIntStubsFcn.h
  /usr/local/include/mysqlStubs.h
  /usr/local/include/odbcStubs.h
  /usr/local/include/pqStubs.h
  /usr/local/include/tcl.h
  /usr/local/include/tclDecls.h
  /usr/local/include/tclOO.h
  /usr/local/include/tclOODecls.h
  /usr/local/include/tclPlatDecls.h
  /usr/local/include/tclThread.h
  /usr/local/include/tclTomMath.h
  /usr/local/include/tclTomMathDecls.h
  /usr/local/include/tdbc.h
  /usr/local/include/tdbcDecls.h
  /usr/local/include/tdbcInt.h
  /usr/local/include/tk.h
  /usr/local/include/tkDecls.h
  /usr/local/include/tkPlatDecls.h

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  ctags
  docutils
  tcl-tk
  remake
```

### Description of issue

A general question: Is there a way to figure out what formulas and casks have recently been added to homebrew? In the past I used `HOMEBREW_UPDATE_REPORT_ALL_FORMULAE` which seemed to tell me but, well:
```
Warning: Calling HOMEBREW_UPDATE_REPORT_ALL_FORMULAE is deprecated! There is no replacement.
Warning: This will not report all formulae because Homebrew cannot get this data from the API.
```

I really liked that feature because I could look at the new formulas/casks that `brew update` reported and discover cool new tools! 

Is there some other command or RSS feed or something I could use to find this info?

**Comment by SMillerDev** [comment]
You can add this as an RSS feed AFAIK: 
- https://github.com/Homebrew/homebrew-core/commits/master
- https://github.com/Homebrew/homebrew-cask/commits/master

**Reply by gromgit** [reply]
I think that should be:
https://github.com/Homebrew/homebrew-core/commits/master.atom
https://github.com/Homebrew/homebrew-cask/commits/master.atom

**Reply by mathomp4** [reply]
Hmm. Can you add a label to these feeds? Like "atom feed with label 'new formula'"?

**Reply by gromgit** [reply]
Not that I know of.

**Comment by chrisfinazzo** [comment]
@mathomp4, Was this always set via an environment variable? Don't recall exactly and only know that it was the default behavior until a point release fairly recently - within the past year, if my memory isn't total garbage about when this happened.

If this isn't a brew-related issue, I will file [elsewhere][], but to the specific case of kegs lacking a formula, I have just come across this with `pycparser` - problem is, this is a dependency for `ocrmypdf` - does anyone know of a replacement we can use instead?

[elsewhere]: https://github.com/ocrmypdf/OCRmyPDF/issues

**Comment by CatGreen90** [comment]
``brew update`` is *supposed* to show new packages (as of release 4.0.5), but for me it only shows deleted packages (including the original names of renamed ones) as new.

**Reply by mathomp4** [reply]
I want to say this all changed with the move to JSON: https://brew.sh/2023/02/16/homebrew-4.0.0/

Maybe there is just no way to know what has changed anymore? 😞

**Reply by mathomp4** [reply]
Maybe I should invoke the name of @MikeMcQuaid and see if there might be a way to get this info (back)? Or maybe the days of discovering new software from `brew update` is just behind us...

**Reply by MikeMcQuaid** [reply]
> I want to say this all changed with the move to JSON: https://brew.sh/2023/02/16/homebrew-4.0.0/

It did.

> Warning: Calling HOMEBREW_UPDATE_REPORT_ALL_FORMULAE is deprecated! There is no replacement.

It's not coming back because this listed all formulae that changed in Git and we no longer have that information in the API.

You should be able to see added/deleted (but not yet renamed, that's WIP) casks/formulae now. If you can't: that's a bug: please file it on Homebrew/brew.

**Reply by mathomp4** [reply]
@MikeMcQuaid Hmm. Interesting. Well, this is on my home laptop, so I'll look at it when I can. Maybe I've defined some weird extra environment variable I don't remember doing. That's my development testbed, so it gets a wacky environment at times from "Try this to fix bug"... 😄

**Reply by MikeMcQuaid** [reply]
@mathomp4 The way the new functionality was implemented: after the fix landed (and you ended up on a tag containing it): the next `brew update` wouldn't show anything new (the files we rely on haven't been populated until this point) but the one after it should be. If not: 🐛 and please report! Thanks for understanding ❤️

**Reply by muescha** [reply]
FYI: the related fix for this bug: https://github.com/Homebrew/brew/pull/15038

**Reply by mathomp4** [reply]
Well, I cleared out any brew related environment variables and things seem...better. Though I still haven't seen any new formulae, I at least saw new casks! Might have just not been any new formulae in the last day or two!

**Reply by mathomp4** [reply]
And I saw new formulae yesterday! I think that means all is well. So I guess it's a combination of cleaning up environment and https://github.com/Homebrew/brew/pull/15038 filtering down to me?

**Reply by MikeMcQuaid** [reply]
Yeh, I also have seen new formulae/casks reported today so seems to be resolved.

**Comment by muescha** [comment]
I also like to discover new formula and casks. My way was to look at the output of `brew update` for formula names and then past it to `brew home {list of names}` to look at each :)

while filing a bug I find this searches for casks useful to show the latest additions and deletions:

- removed casks: https://github.com/Homebrew/homebrew-cask/pulls?q=is%3Apr+sort%3Aupdated-desc+remove+is%3Aclosed+label%3A%22marked+for+removal%2Frejection%22
- new casks: https://github.com/Homebrew/homebrew-cask/pulls?q=is%3Apr+sort%3Aupdated-desc+label%3A%22new+cask%22+is%3Amerged+

but i would also like to see a info stream about additions, deletions and modifications

**Reply by MikeMcQuaid** [reply]
@muescha note: modifications will not be implemented, we cannot obtain that information without Git.

**Comment by tyrann0us** [comment]
BTW, if anyone is interested, I asked for the opposite feature in https://github.com/orgs/Homebrew/discussions/4590; a way to disable the output of the "New Formulae"/"New Casks" ~and "Deleted Formulae"/"Deleted Casks"~ sections when running `brew update`.

**Reply by chrisfinazzo** [reply]
Definitely - In the case of "Deleted", I would argue the only time you _really_ need to see that is if one of your current installs is dropped from the main repo.

In that case, I'd hope that whoever maintains the affected formula or cask would tell you how to proceed - be it a tap or an outright replacement.

**Comment by marcotrosi** [comment]
I also want the feature back, I want to learn about new formulas the easy way.
Currently I use an alias to run `git -C /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core log --pretty=format:%s --since=2 weeks ago` followed by some grepping and text manipulation, but it's annoying to run it manually and I have to remember which tools I already have seen as new. I don't understand why it was decided to remove such a useful feature.

**Reply by SMillerDev** [reply]
Because [homebrew is not a discoverability service](https://docs.brew.sh/Acceptable-Casks#homebrew-cask-is-not-a-discoverability-service) and the choice was made to give people a significantly faster and smaller install with smoother updates instead of keeping this thing that Homebrew never meant to support. 

If you want to keep track of all new things in Homebrew you can follow the rss feed for the repo. That'll allow you to mark items as read too.

**Reply by muescha** [reply]
the feature request in the linked section is about a discovery service (for example group by category and more) by Homebrew itself. 

but here is a request to get back a previous existing function "Information about (new|update|delete) of a (formula|casks)". these informations already exists while processing the cask meta data - for example for the updated formula/casks otherwise brew outdated not work

**Reply by muescha** [reply]
> Currently I use an alias to run `git -C ...`

since change to the new api there is no git repo `homebrew-core` under `/opt/homebrew/Library/Taps/homebrew` - so your alias not works :(

**Reply by SMillerDev** [reply]
And the ones that are available with the API have been restored. The other ones never will because it requires giving up all the improvements the API brings for the gain of a feature Homebrew never intended to provide.

**Reply by osalbahr** [reply]
> > Currently I use an alias to run `git -C ...`
> 
> since change to the new api there is no git repo `homebrew-core` under `/opt/homebrew/Library/Taps/homebrew` - so your alias not works :(

FYI, you can still `brew tap homebrew/core` if you want. Note that it will take a long time and take up some storage. You might need the `--force` flag.

**Reply by muescha** [reply]
oh missed that the report of new and updated formula/casks are back :)

see:
- https://github.com/Homebrew/brew/pull/14802
- https://brew.sh/2023/07/20/homebrew-4.1.0/

PS:
somehow i still was on 4.0.10 - courious, because some days ago (2023-07-30) I got an update of the brew system (with the `brew update --auto-update`) - and I assumed that I am on edge because some files are changed - but it was only an update to 4.0.10? 

now I am on 4.1.3 after an `brew update` :)

**Reply by marcotrosi** [reply]
<img width="866" alt="image" src="https://github.com/Homebrew/discussions/assets/8596795/1b62e991-2b07-423b-a943-1b62143ea340">

my alias still delivers the new packages. I dont know if the output is correct, but at least it doesnt fail.
